src/processor/operator/copy_to/csv_file_writer.cpp

```diff
@@ -0,0 +1,127 @@
+    auto value = TypeUtils::toString(vector->getValue<T>(pos));
+    if (bindStringValue) {
+        bindString(value);
+    }
+    addToBuffer(value);
+}
+
+void CSVFileWriter::open(const std::string& filePath) {
+    fileInfo = common::FileUtils::openFile(filePath, O_WRONLY | O_CREAT | O_TRUNC);
+}
+
+void CSVFileWriter::writeHeader() {
+    auto columns = getColumnNames();
+    auto columnsSize = columns.size();
+    if (columnsSize > 0) {
+        addToBuffer(columns[0]);
+        for (uint64_t i = 1; i < columnsSize; i++) {
+            addToBuffer(DELIMITER);
+            addToBuffer(columns[i]);
+        }
+        addToBuffer(LINE_BREAK);
+        flush();
+    }
+}
+
+void CSVFileWriter::flush() {
+    const std::string str = buffer.str();
+    FileUtils::writeToFile(fileInfo.get(), (uint8_t*)str.data(), str.size(), fileOffset);
+    fileOffset += str.size();
+    buffer.str("");
+}
+
+void CSVFileWriter::bindString(std::string& value) {
+    StringUtils::replaceAll(value, "\"", "\"\"");
+    value = "\"" + value + "\"";
+}
+
+void CSVFileWriter::write(common::ValueVector* vector, int64_t pos) {
+    auto position = vector->state->selVector->selectedPositions[pos];
+    switch (vector->dataType.getLogicalTypeID()) {
+    case LogicalTypeID::BOOL:
+        return addToBuffer<int8_t>(vector, position);
+    case LogicalTypeID::INT64:
+        return addToBuffer<int64_t>(vector, position);
+    case LogicalTypeID::INT32:
+        return addToBuffer<int32_t>(vector, position);
+    case LogicalTypeID::INT16:
+        return addToBuffer<int16_t>(vector, position);
+    case LogicalTypeID::DOUBLE:
+        return addToBuffer<double>(vector, position);
+    case LogicalTypeID::FLOAT:
+        return addToBuffer<float>(vector, position);
+    case LogicalTypeID::DATE:
+        return addToBuffer<date_t>(vector, position, true);
+    case LogicalTypeID::TIMESTAMP:
+        return addToBuffer<timestamp_t>(vector, position, true);
+    case LogicalTypeID::INTERVAL:
+        return addToBuffer<interval_t>(vector, position, true);
+    case LogicalTypeID::STRING:
+        return addToBuffer<ku_string_t>(vector, position, true);
+    case LogicalTypeID::INTERNAL_ID:
+        return addToBuffer<internalID_t>(vector, position, true);
+    case LogicalTypeID::VAR_LIST:
+    case LogicalTypeID::FIXED_LIST:
+        return addListToBuffer<list_entry_t>(vector, position);
+    case LogicalTypeID::STRUCT:
+        return addListToBuffer<struct_entry_t>(vector, position);
+    default: {
+        // suppress warning
+    }
+    }
+}
+
+void CSVFileWriter::run(std::vector<common::ValueVector*>& outputVectors) {
+    bool hasData = true;
+    int64_t pos = 0;
+    auto outputVectorsSize = outputVectors.size();
+    if (outputVectorsSize == 0) {
+        return;
+    }
+    // 'pos' represents the position of the row
+    // 'vectorPos' represents the position of the column
+    while (hasData) {
```

Comment:
As discussed, we should assume all flat input for now. So I don't the logic should be complicated as the one you have.
Generally, it should follow something like this:
```
func writeTuple() {
    for (auto vector : vectors) {
        writeValue(vector)
    }
}

func writeValue(vector) {
    assert(vector is flat)
    write value at flatten pos. selectedPos[0]
}
```